train_segmentation.py

```diff
@@ -306,15 +306,15 @@ def main(params, config_path):
                                                                        samples_folder=samples_folder)
 
     tqdm.write(f'Setting model, criterion, optimizer and learning rate scheduler...\n')
-    model, criterion, optimizer, lr_scheduler = set_hyperparameters(params, num_classes_corrected, model, checkpoint)
-
-    criterion = criterion.to(device)
-    try: # For HPC when device 0 not available. Error: Cuda invalid device ordinal.
+    try:  # For HPC when device 0 not available. Error: Cuda invalid device ordinal.
         model.to(device)
     except RuntimeError:
         warnings.warn(f"Unable to use device. Trying device 0...\n")
         device = torch.device(f'cuda:0' if torch.cuda.is_available() and lst_device_ids else 'cpu')
         model.to(device)
+    model, criterion, optimizer, lr_scheduler = set_hyperparameters(params, num_classes_corrected, model, checkpoint)
+
+    criterion = criterion.to(device)
```

Comment:
I don't know if it has changed or not in pytorch but as I recall, we had to initialize the optimizer before pushing the model to cuda device. @remtav do you recall this as well?  
Here's an explanation: https://discuss.pytorch.org/t/effect-of-calling-model-cuda-after-constructing-an-optimizer/15165

Reply:
I encountered an error when initalizing the optimizer before pushing the model to cuda. Construting the initializer uses the model parameters as seen from the code. However when it is constructed before the model is pushed for instance to a cuda device(GPU) a mismatch occurs between the parameters ( model initialized on CPU vs Model pushed to Cuda device).  Apparently, this problem does not occur for the SGD optimizer, but for ADAM and AdaGrad the problem is encountered. Here are some official docs and discussion on the issue:
https://pytorch.org/docs/master/optim.html
https://discuss.pytorch.org/t/effect-of-calling-model-cuda-after-constructing-an-optimizer/15165

Reply:
Well the doc seems clear: better to construct the optimizer after pushing model to GPU. I seem to recall this part of the code was a little tricky to move and we had to do everything in a particular order. But anyways, I does seem more logical with the provided doc to push model before constructing optimizer...